util/coinach.ts

```diff
@@ -0,0 +1,176 @@
+  }
+
+  async exd(table: string, lang: string): Promise<string[]> {
+    return await this._coinachCmd('exd', table, lang);
+  }
+  async rawexd(table: string, lang: string): Promise<string[]> {
+    return await this._coinachCmd('rawexd', table, lang);
+  }
+  async _coinachCmd(coinachCmd: string, table: string, lang: string): Promise<string[]> {
+    const cmdList = [path.join(this.coinachPath, _COINACH_EXE)];
+    const args = [this.ffxivPath, 'lang ' + lang, coinachCmd + ' ' + table];
+
+    const cmd = [...cmdList, ...args].map((x) => `"${x}"`).join(' ');
+
+    if (this.verbose) {
+      console.log(`coinach_path: ${this.coinachPath}`);
+      console.log('ffxiv_path: ' + this.ffxivPath);
+      console.log('cmd: ' + cmd);
+    }
+    // # This will throw an exception if stuff is VERY wrong
+    // # however, return code is still 0 even if all exports fail.
+    // # Also, it seems to need to be run from the SaintCoinach directory.
+    const output = String(promisify(exec)(cmd));
+    // # Manually check output for errors.
+    let m = /^([0-9])* files exported, ([0-9])* failed/m.exec(output);
+    if (!m) throw new CoinachError('Unknown output', cmd, output);
+    if (m[1] === '0')
+      throw new CoinachError('Zero successes', cmd, output);
+    if (m[2] !== '0')
+      throw new CoinachError('Non-zero failures', cmd, output);
+
+    // # Find directory that this export was written to.
+    // # There's no way to control this.
+    m = /^Definition version: ([0-9.]*)/m.exec(output);
+    if (!m) throw new CoinachError('Unknown output', cmd, output);
+
+    const csvFilename = path.join(
+        this.coinachPath,
+        m[1] ?? '',
+        coinachCmd,
+        `${table}.csv`,
+    );
+
+    if (this.verbose) {
+      console.log(`output: \noutput`);
+      console.log(`output csv: ${csvFilename}`);
+    }
+    // # Read the whole file immediately,
+    // # as future commands with different langs will overwrite.
+    const lines = String(await promisify(fs.readFile)(csvFilename, {
+      flag: 'r',
+      encoding: 'utf-8',
+    })).split(/\r?\n/);
+
+    if (this.verbose) console.log(`csv lines: ${lines.length}`);
+
+    return lines;
+  }
+}
+
+export class CoinachWriter {
+  cactbotPath: string;
+  verbose: boolean;
+  constructor(cactbotPath: string | null, verbose: boolean) {
+    this.verbose = verbose;
+
+    if (!cactbotPath) cactbotPath = this._findCactbotPath();
+    this.cactbotPath = cactbotPath;
+    if (!fs.existsSync(this.cactbotPath))
+      throw new Error(`Invalid cactboth path: ${this.cactbotPath}`);
+  }
+  _findCactbotPath(): string {
+    return '.';
+  }
+  write(filename: string, scriptname: string, variable: string | null, d: unknown[]): void {
+    const fullPath = path.join(this.cactbotPath, filename);
+
+    let f = `// Auto-generated from ${scriptname}
+    // DO NOT EDIT THIS FILE DIRECTLY
+
+
+    export default `;
+
+    let str = JSON.stringify(d.sort());
```

Comment:
`JSON.stringify` is not the same as `json.dumps(d, sort_keys=True, indent=2, ensure_ascii=False)`.  In particular, it's missing the indent=2 and `sort` isn't a function on objects.

Re: indenting, one possibility here is to run eslint, see: https://github.com/quisquous/cactbot/blob/main/util/gen_coverage_report.js#L230-L248

Re: sorting, I think `JSON.stringify(d, Object.keys(d).sort())` will sort the top level keys.  If that's not enough and we need to sort the inner keys too, maybe we need https://www.npmjs.com/package/json-stable-stringify?

Reply:
[JSON.stringify()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) can take a third argument for spacing; I don't know if that covers the need here; `JSON.stringify(d, Object.keys(d).sort(), 2)` in this instance.

Reply:
Ah yeah, thanks! I think we'd still want to eslint as well, though.